`OSSL_CMP_SRV_process_request()`: fix detection of new transactions and the context re-initialization

DDvO opened the pull request:
* fix wrong handling of aborted polling transaction
  as well as the related test, where missing parentheses prevented reporting failure.
  
  The check for acceptable response body types during polling
  should only be done when we are in an active transaction.
* make detection of new transactions and the context re-initialization more complete.
  This motivated factoring out new functions: `assuming_new_transaction()` and `transaction_reinit()`.
  
  This includes a minor fix:
  So far, receiving an error message from the client was taken is an indication for a new transaction, 
  which does not make much sense. So this case has been dropped in assuming_new_transaction().

DDvO committed:
OSSL_CMP_SRV_process_request(): make detection of new transactions and the context re-initialization more complete

This motivated factoring out new functions: assuming_new_transaction() and transaction_reinit()

This includes a minor fix:
So far, receiving an error message from the client was taken is an indication for a new transaction,
which does not make much sense. So this case has been dropped in assuming_new_transaction().

DDvO committed:
OSSL_CMP_SRV_process_request(): fix wrong handling of aborted polling transaction

The check for acceptable response body types during polling
should only be done when we are in an active transaction.

Also fix missing parentheses in est/cmp_client_test.c, which had prevented reporting failure.

DDvO commented:
CI failure (apparently a stuck run, which I experienced recently also with another PR) is unrelated.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

DDvO commented:
Ping @openssl/committers for reviews

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

esyr commented on crypto/cmp/cmp_server.c:
```diff
@@ -654,12 +654,6 @@ OSSL_CMP_MSG *OSSL_CMP_SRV_process_request(OSSL_CMP_SRV_CTX *srv_ctx,
     if (!OSSL_CMP_CTX_set1_recipient(ctx, hdr->sender->d.directoryName))
```
Would be nice to have the commit message a bit more elaborate.

DDvO replied:
added, also to the OP:
```
The check for acceptable response body types during polling
should only be done when we are in an active transaction.
```

esyr commented on crypto/cmp/cmp_server.c:
```diff
@@ -571,6 +570,54 @@ static int unprotected_exception(const OSSL_CMP_CTX *ctx,
+    case OSSL_CMP_PKIBODY_P10CR:
+    case OSSL_CMP_PKIBODY_KUR:
+    case OSSL_CMP_PKIBODY_RR:
+    case OSSL_CMP_PKIBODY_GENM:
```
`OSSL_CMP_PKIBODY_ERROR` is no longer there, would be nice to mention why, in the commit message, for example.

DDvO replied:
Added to the commit message and OP:
```
This includes a minor fix:
So far, receiving an error message from the client was taken is an indication for a new transaction,
which does not make much sense. So this case has been dropped in assuming_new_transaction().
```

esyr commented on crypto/cmp/cmp_server.c:
```diff
@@ -571,6 +570,54 @@ static int unprotected_exception(const OSSL_CMP_CTX *ctx,
+    }
+
+    switch (body_type) {
+    case OSSL_CMP_PKIBODY_IR:
```
I'm trying to cross-reference it with the types described in [1] and wonder why `popdecc`, `krr`, `ccr`, `ckuann`, `cann`, and `rann` messages are not here.

[1] https://www.rfc-editor.org/info/rfc9810/#name-pki-message-body

DDvO replied:
This is simply because those CMP message body types are not supported.
The ones we support are those specified in the [Lightweight CMP Profile [RFC 9483]](https://www.rfc-editor.org/info/rfc9483/).
RFC 4210, and thus also its successor RFC 9810, contains various further very uncommon CMP message types, including those you mentioned.

esyr commented on crypto/cmp/cmp_server.c:
```diff
@@ -571,6 +570,54 @@ static int unprotected_exception(const OSSL_CMP_CTX *ctx,
+    return OSSL_CMP_CTX_set1_transactionID(srv_ctx->ctx, NULL)
+        && OSSL_CMP_CTX_set1_senderNonce(srv_ctx->ctx, NULL)
+        && (srv_ctx->clean_transaction == NULL
+            || srv_ctx->clean_transaction(srv_ctx, srv_ctx->ctx->transactionID));
```
Wouldn't `transactionID` be `NULL` after `OSSL_CMP_CTX_set1_transactionID(srv_ctx->ctx, NULL)`? Looks like it was passed before resetting in response handling code.

DDvO replied:
Right, `OSSL_CMP_CTX_set1_transactionID(srv_ctx->ctx, NULL)` sets `srv_ctx->ctx->transactionID` to `NULL`.
This is done here in order to become open for a new transaction at the end of the current transaction.
The only other place where this field is set at the server side is in `ossl_cmp_msg_check_update()` of `cmp_vfy.c`, where the server learns at the (non-NULL) new transactionID of the first/next transaction, which was sent by a client.

andrewkdinh replied:
@DDvO This is still an issue

`OSSL_CMP_CTX_set1_transactionID(srv_ctx->ctx, NULL)` runs first and clears `ctx->transactionID`, so by the time `clean_transaction(srv_ctx, srv_ctx->ctx->transactionID)` is evaluated, it always receives `NULL`.

So the fix would be to capture `srv_ctx->ctx->transactionID` before then pass it as a parameter after

DDvO replied:
Ah, now I understand what @esyr meant here! Good catch.
Thanks @andrewkdinh for explaining.

I fixed this by simply returning to the original order: first call `clean_transaction(srv_ctx, srv_ctx->ctx->transactionID)` (if applicable), and then call `OSSL_CMP_CTX_set1_transactionID(srv_ctx->ctx, NULL)`.

This avoids the imminent regression you spotted.

esyr commented on test/cmp_client_test.c:
```diff
@@ -109,9 +109,7 @@ static int execute_exec_GENM_ses_test_single(CMP_SES_TEST_FIXTURE *fixture)
-            && fixture->expected == OSSL_CMP_PKISTATUS_accepted
-        ? TEST_ptr(itavs)
-        : TEST_ptr_null(itavs);
+        && (fixture->expected == OSSL_CMP_PKISTATUS_accepted ? TEST_ptr(itavs) : TEST_ptr_null(itavs));
```
Would be nice to have a more elaborate description in the commit message that mentions it.

DDvO replied:
added:
`Also fix missing parentheses in est/cmp_client_test.c, which had prevented reporting failure.`

esyr reviewed (changes requested):
I suspect that resetting `transactionID` before calling the `clean_transaction` callback is a mistake, please check (and maybe add some test coverage for it).

DDvO commented:
Thank you @esyr for reviewing this PR and the comments resulting from this!
I've just handled them, resulting in extended commit messages (only).

DDvO commented:
@esyr did you see that I handled your review comments?

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

DDvO committed:
fixup! OSSL_CMP_SRV_process_request(): make detection of new transactions and the context re-initialization more complete

andrewkdinh commented on crypto/cmp/cmp_server.c:
```diff
@@ -571,6 +570,54 @@ static int unprotected_exception(const OSSL_CMP_CTX *ctx,
+    tid = OSSL_CMP_HDR_get0_transactionID(OSSL_CMP_MSG_get0_header(req));
+    if (tid != NULL && ASN1_OCTET_STRING_cmp(tid, ctx->transactionID) != 0) {
+        char *ctx_str = i2s_ASN1_OCTET_STRING(NULL, ctx->transactionID);
+        char *tid_str = i2s_ASN1_OCTET_STRING(NULL, tid);
```
These can return NULL

DDvO replied:
I've meanwhile added case distinctions, to be on the safe side.

andrewkdinh commented on crypto/cmp/cmp_server.c:
```diff
@@ -571,6 +570,54 @@ static int unprotected_exception(const OSSL_CMP_CTX *ctx,
+    return ((srv_ctx->clean_transaction == NULL
+                || srv_ctx->clean_transaction(srv_ctx, srv_ctx->ctx->transactionID))
+        && OSSL_CMP_CTX_set1_transactionID(srv_ctx->ctx, NULL)
+        && OSSL_CMP_CTX_set1_senderNonce(srv_ctx->ctx, NULL));
```
These 2 used to be called unconditionally, but they won't be called if `srv_ctx->clean_transaction()` fails

DDvO replied:
Good point - fixed this aspect as well.

DDvO commented:
While handling the most recent review comments, as well as new Copilot comments,
I noticed that some documentation extensions were due, which I added,
and rebased the PR again to fix a new merge conflict.
On this occasion also added, in a separate commit, doc references to the 
Lightweight CMP Profile, which is the focus of this CMP implementation.

DDvO commented:
Thanks. @andrewkdinh for your comments and first approval.
@esyr did you see that I handled also your comments?
Could you please have a look again and hopefully approve as well.

DDvO commented:
Ping @openssl/committers

esyr commented on doc/man3/OSSL_CMP_SRV_CTX_new.pod:
```diff
@@ -94,15 +94,22 @@ OSSL_CMP_SRV_CTX_set_grant_implicit_confirm
 
```
Nits regarding the commit message: ideally, subject should fit into 80 characters (having something like `doc/man3/OSSL_CMP_{exec_certreq,SRV_CTX}_new.pod: enrich CMP client/server doc`, maybe? Or, just drop the `of CMP client and server` part?);  commit message itself is also should be wrapped to 72—80 character line width, ideally.

Also, `s/singel/single/`

DDvO replied:
I've shortened the the commit header to:
`OSSL_CMP_{exec_certreq,SRV_CTX}_new.pod: enrich doc of CMP client and server`
and fixed the typo in the commit message.

esyr commented on doc/man3/OSSL_CMP_SRV_CTX_new.pod:
```diff
@@ -94,15 +94,22 @@ OSSL_CMP_SRV_CTX_set_grant_implicit_confirm
 I<OSSL_CMP_CTX_get_status(OSSL_CMP_SRV_CTX_get0_cmp_ctx(srv_ctx))> yields -1
 then the function has closed the current transaction,
 which may be due to normal successful end of the transaction or due to an error.
+Otherwise it should be called again with the next request message of the same
```
`s/Otherwise/Otherwise,/`

esyr commented on doc/man3/OSSL_CMP_SRV_CTX_new.pod:
```diff
@@ -94,15 +94,22 @@ OSSL_CMP_SRV_CTX_set_grant_implicit_confirm
 then the function has closed the current transaction,
 which may be due to normal successful end of the transaction or due to an error.
+Otherwise it should be called again with the next request message of the same
+transaction (which can be a certificate confirmation or poll request message).
```
`s/poll/a poll/`

esyr commented on doc/man3/OSSL_CMP_SRV_CTX_new.pod:
```diff
@@ -94,15 +94,22 @@ OSSL_CMP_SRV_CTX_set_grant_implicit_confirm
+Otherwise it should be called again with the next request message of the same
+transaction (which can be a certificate confirmation or poll request message).
+The function should not be called by multiple threads sharing
+the same I<srv_ctx>; calling it in parallel leads to undefined behavior.
```
"without explicitly serialising the calls"?

esyr commented on doc/man3/OSSL_CMP_SRV_CTX_new.pod:
```diff
@@ -160,11 +167,21 @@ confirmation of newly enrolled certificates if requested.
-(and consequently to at most one response component per CMP message).
+Like the OpenSSL CMP client, the CMP server implementation documented here
+focuses on the Lightweight CMP Profile (RFC 9483).
+Among others, this implies that only commonly used CMP message types are supported
```
"Among other things"? Or drop the adverbial phrase altogether.

esyr commented on doc/man3/OSSL_CMP_SRV_CTX_new.pod:
```diff
@@ -160,11 +167,21 @@ confirmation of newly enrolled certificates if requested.
+
+So far, this server implementation is single-threaded and
+should not be called in parallel for any given B<OSSL_CMP_SRV_CTX> I<srv_ctx>.
+It can handle only one CMP transaction at a time (which of course is not a problem
```
`s/ of course/, of course,/

esyr commented on doc/man3/OSSL_CMP_SRV_CTX_new.pod:
```diff
@@ -94,15 +94,22 @@ OSSL_CMP_SRV_CTX_set_grant_implicit_confirm
 
-OSSL_CMP_SRV_process_request() implements the generic aspects of a CMP server.
-Its arguments are the B<OSSL_CMP_SRV_CTX> I<srv_ctx> and the CMP request message
+OSSL_CMP_SRV_process_request() implements the generic core aspects of a CMP server.
```
Hm, it might be worth expanding the CMP acronym on the first usage here as well, similarly to how it is done in `OSSL_CMP_exec_certreq.pod`.

esyr commented on doc/man3/OSSL_CMP_exec_certreq.pod:
```diff
@@ -51,8 +51,9 @@ OSSL_CMP_get1_certReqTemplate
 
-This is the OpenSSL API for doing CMP (Certificate Management Protocol)
-client-server transactions, i.e., sequences of CMP requests and responses.
+This is the OpenSSL main API for CMP (Certificate Management Protocol) clients.
```
Stylistically, I prefer putting acronym in the parentheses and not the expansion, but won't insist, as OpenSSL doesn't have any established style guide.

esyr commented on doc/man3/OSSL_CMP_exec_certreq.pod:
```diff
@@ -185,8 +186,9 @@ Both must be freed by the caller.
-The CMP client implementation is limited to one request per CMP message
-(and consequently to at most one response component per CMP message).
+This CMP implementation focuses on the Lightweight CMP Profile (RFC 9483).
+Among others, this implies that only commonly used CMP message types are supported
```
"Among other things"?

esyr commented on crypto/cmp/cmp_server.c:
```diff
@@ -55,7 +55,6 @@ OSSL_CMP_SRV_CTX *OSSL_CMP_SRV_CTX_new(OSSL_LIB_CTX *libctx, const char *propq)
     if ((ctx->ctx = OSSL_CMP_CTX_new(libctx, propq)) == NULL)
```
Nit: commit message subject should be under 80 characters, ideally (but I can't come up quickly with something more laconic, so, doesn't matter much).

There's a typo in the commit message: `s/is an/as an/`

I'd suggest to reformat the commit message into something like
```
This motivated factoring out new functions: assuming_new_transaction()
and transaction_reinit().  It also includes a minor fix: so far, receiving
an error message from the client was taken as an indication for a new
transaction, which does not make much sense;  so, this case has been dropped
in assuming_new_transaction().
```
, but you can ignore it.

DDvO replied:
I've somewhat shortened the the commit header to:
`OSSL_CMP_SRV_process_request(): complete transaction detection and context re-initialization
`
and re-phrased and re-formatted the commit message as you suggested.

esyr commented on crypto/cmp/cmp_server.c:
```diff
@@ -654,12 +654,6 @@ OSSL_CMP_MSG *OSSL_CMP_SRV_process_request(OSSL_CMP_SRV_CTX *srv_ctx,
     if (!OSSL_CMP_CTX_set1_recipient(ctx, hdr->sender->d.directoryName))
```
There are a missing comma and a typo ("est") in the last line of the commit message:
```
Also, fix missing parentheses in test/cmp_client_test.c, which
had prevented reporting failure.
```

esyr commented:
> @esyr did you see that I handled also your comments? Could you please have a look again and hopefully approve as well.

My apologies, I dropped the ball wrt PR reviews recently, unfortunately.

DDvO commented:
Merged to `master`, 4.0, 3.6, and 3.5.
Thanks again @andrewkdinh for your review and approval,
and thanks @esyr for picking up the ball again and your final round of (minor) comments.
I've done all the doc and merge message nit fixes, as suggested during the merge.
